tket/src/WeightSubgrMono/GraphTheoretic/VertexRelabelling.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/GraphTheoretic/VertexRelabelling.hpp"
+
+#include <sstream>
+#include <stdexcept>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+VertexRelabelling::VertexRelabelling(GraphEdgeWeights edges_and_weights)
+    : new_edges_and_weights(std::move(edges_and_weights)) {
+  for (const auto& entry : new_edges_and_weights) {
+    const auto& v1 = entry.first.first;
+    const auto& v2 = entry.first.second;
+    if (v1 == v2) {
+      throw std::runtime_error(
+          std::string("Loop found in graph; not allowed: v=") +
+          std::to_string(v1));
+    }
+    old_to_new_vertex_labels[v1];
+    old_to_new_vertex_labels[v2];
+    const auto weight_opt =
+        get_optional_value(new_edges_and_weights, std::make_pair(v2, v1));
+    if (weight_opt) {
+      const auto other_weight = weight_opt.value();
+      if (entry.second != other_weight) {
+        std::stringstream ss;
+        ss << "Edge (" << v1 << "," << v2 << ") has weight " << entry.second
+           << ", differs from edge (" << v2 << "," << v1 << ") with weight "
+           << other_weight;
+        throw std::runtime_error(ss.str());
+      }
+    }
+  }
+  if (old_to_new_vertex_labels.empty()) {
+    throw std::runtime_error("no edges!");
+  }
+  number_of_vertices = old_to_new_vertex_labels.size();
+  TKET_ASSERT(number_of_vertices >= 2);
+  if (old_to_new_vertex_labels.cbegin()->first == 0 &&
+      old_to_new_vertex_labels.crbegin()->first == number_of_vertices - 1) {
+    // Already contiguous!
+    old_to_new_vertex_labels.clear();
+    return;
+  }
+  new_to_old_vertex_labels.reserve(number_of_vertices);
+  for (auto& entry : old_to_new_vertex_labels) {
+    const auto new_vertex = new_to_old_vertex_labels.size();
+    new_to_old_vertex_labels.push_back(entry.first);
+    entry.second = new_vertex;
+  }
+  TKET_ASSERT(new_to_old_vertex_labels.size() == number_of_vertices);
+  TKET_ASSERT(old_to_new_vertex_labels.at(new_to_old_vertex_labels[0]) == 0);
```

Comment:
Ah, this is a case of a cheap assert, rather than a full assert!

Reply:
(This exact assert DID catch a bug when I was first writing the code, which DID save me at least a few minutes; I am reluctant to remove them because who knows when things will go wrong again in future?)

Reply:
What was the bug it caught before and how did it happen? Have you changed the loop since?

Reply:
Sorry, I really think you're missing the point. When I write code initially, I put asserts in, and some of them catch bugs, saving me debugging time.

Obviously, now that I've corrected the bugs, they are (presumably) no longer present (if a bug is still present, it must be very subtle) - but we should still leave the asserts IN, who knows when they may prove useful later?

It may seem like a minor issue, but I very strongly feel that I am right in this case.